core/trino-spi/src/main/java/io/trino/spi/cache/CacheManager.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.cache;
+
+import io.trino.spi.connector.ConnectorPageSink;
+import io.trino.spi.connector.ConnectorPageSource;
+import io.trino.spi.predicate.TupleDomain;
+
+import java.io.Closeable;
+import java.util.Optional;
+
+public interface CacheManager
+{
+    /**
+     * @return {@link SplitCache} for a given {@link PlanSignature}.
+     * Matching of {@link PlanSignature} per split could be expensive,
+     * therefore {@link SplitCache} is used to load or store data per split.
+     */
+    SplitCache getSplitCache(PlanSignature signature);
+
+    /**
+     * Triggers a memory revoke. {@link CacheManager} should revoke
+     * at least {@code bytesToRevoke} bytes (if it has allocated
+     * that much revocable memory) before allocating new memory.
+     *
+     * @return the number of revoked bytes
+     */
+    long revokeMemory(long bytesToRevoke);
+
+    interface SplitCache
+            extends Closeable
+    {
+        /**
+         * @param predicate Predicate that should be enforced on cached rows.
+         * Output of `cachedSplitA` can be used to derive output of matching `cachedSplitB`
+         * (with corresponding {@link PlanSignature}) as long as `cachedSplitB.predicate` is a strict
+         * subset of `cachedSplitA.predicate`. To do so, `cachedSplitB.predicate` must be
+         * applied on output of `cachedSplitA`. Before serialization as a cache key, predicate
+         * needs to be normalized using {@code io.trino.plugin.base.cache.CacheUtils#normalizeTupleDomain(TupleDomain)}.
```

Comment:
How is this enforced? Who is responsible for the normalization? What happens if it's not normalized?

Reply:
> How is this enforced?

It's not enforced.

>  Who is responsible for the normalization?

The concrete `CacheManager` if it would like to serialize these `TupleDomain`s and use the result as part of its cache key (for instance, `MemoryCacheManager` doesn't do that)

> What happens if it's not normalized?

Equal tuple domains might not be serialized in the same way. A `CacheManager` that serializes these predicates without normalization into its cache key, might experience unnecessary cache misses.